File: cpp/509_fibonacci_number/main.cpp
#include <map>

class Solution
{
public:
    int fib(int n)
    {
        std::map<int, int> mem;
        return fib(n, mem);
    }

    int fib2(int n)
    {
        if (n == 0 || n == 1)
        {
            return n;
        }
        
        int previous = 0;
        int current = 1;
        int i = 2;

        int tmp = 0;
        while (i <= n)
        {
            tmp = previous + current;
            previous = current;
            current = tmp;
            i++;
        }
        return current;
            
    }

private:

    int fib(int n, std::map<int, int>& mem)
    {
        if (n == 0 || n == 1)
        {
            return n;
        }
        
        if (mem.contains(n))
        {
            return mem[n];
        }

        mem[n] = fib(n - 2, mem) + fib(n - 1, mem);
        return mem[n];
    }
};


#include <iostream>

int main(void)
{
    Solution s;
    std::cout << s.fib(4) << "\n";
    std::cout << s.fib2(4) << "\n";    
}

File: ruby/509_fibonacci_number/main.rb
# @param {Integer} n
# @return {Integer}
def fib(n)
  if n == 0 || n == 1
    return n
  end

  previous = 0
  current = 1

  i = 2

  while i <= n
    tmp = previous + current
    previous = current
    current = tmp
    i += 1
  end
  current
end

puts fib(2)
puts fib(3)
puts fib(4)
puts fib(5)
